cmd/earthly/main.go

```diff
@@ -2831,15 +2850,40 @@ func (app *earthlyApp) actionBuildImp(c *cli.Context, flagArgs, nonFlagArgs []st
 		return errors.Errorf("multi-platform builds are not yet supported on the command line. You may, however, create a target with the instruction BUILD --plaform ... --platform ... %s", target)
 	}
 
-	dotEnvMap, err := godotenv.Read(app.envFile)
+	dotEnvMap, err := readEnvFile(app.envFile, app.envFile == defaultEnvFile)
 	if err != nil {
-		// ignore ErrNotExist when using default .env file
-		if app.envFile != defaultEnvFile || !errors.Is(err, os.ErrNotExist) {
-			return errors.Wrapf(err, "read %s", app.envFile)
+		return err
+	}
+	argsMap, err := readEnvFile(app.argsFile, app.argsFile == defaultArgsFile)
+	if err != nil {
+		return err
+	}
+	secretsStringMap, err := readEnvFile(app.secretsFile, app.secretsFile == defaultSecretsFile)
+	if err != nil {
+		return err
+	}
+
+	nonEarthlyEnvs := []string{}
+	for key, val := range dotEnvMap {
+		if !app.isEarthlyFlagEnvName(key) {
+			nonEarthlyEnvs = append(nonEarthlyEnvs, key)
+			if _, found := argsMap[key]; found {
+				return errors.Errorf("build arg %s found in both %s and %s", key, app.envFile, app.argsFile)
+			}
+			if _, found := secretsStringMap[key]; found {
+				return errors.Errorf("secret %s found in both %s and %s", key, app.envFile, app.secretsFile)
+			}
+			argsMap[key] = val
+			secretsStringMap[key] = val
 		}
 	}
+	if len(nonEarthlyEnvs) > 0 {
+		sort.Strings(nonEarthlyEnvs)
+		keys := strings.Join(nonEarthlyEnvs, ", ")
+		app.console.Warnf("Warning: build-args or secrets (%s) should be defined in .earthlyargs or .earthlysecrets; this will become mandatory in a future version of earthly.", keys)
```

Comment:
If we decide to GA this later, we'd need to take out this message and re-add it later (e.g. in v0.7).

Reply:
I added a `hideEarthlySecretsAndArgs = true` const to tie them all together.